### PR TITLE
Support for project level .jshintignore

### PR DESCRIPTION
**Pull request by omarestrella**
(Opening changes since last PR, #40. That one closed on me while I was cleaning up.)

This respects a `.jshintignore` file in a project and does not process strings to generate the tests for jshint.

### CODE REVIEW

**Comment by omarestrella** [comment]
>@omarestrella I think this is heading the right direction, but I would instead want this to build a new broccoli node (tree in old parlance) that it can blindly process rather than doing a repeated ignoredFiles check.

@nathanhammond I investigated this and couldn't figure out how to get a broccoli plugin/node to call into another broccoli plugin/node. I imagine it would be find inside of ember-cli/brocfiles by merging the trees, but thats not an avenue here, correct? If I am missing something, could you point me in the right path (documentation, code) to get started on what you are looking for?

Ultimately, I did fine that there is a public method called `getDestFilePath` that mentioned:
>Virtual method `getDestFilePath`: determine whether the source file should
>be processed, and optionally rename the output file when processing occurs.
>
>Return `null` to pass the file through without processing. Return
>`relativePath` to process the file with `processString`. Return a
>different path to process the file with `processString` and rename it.

Would this be a place to hook into?

**Comment by nathanhammond** [comment]
So the idea is to add the entire root node to the project and then run it through a filter to remove files identified by `.jshintignore`. You would do this prior to any of the current jshint operations.

More explicitly:
1. The input node is the entire application dir tree.
2. You then run that tree through a filter of your own devising that is basically `files.some(checkIgnored)`. This returns a new tree.
3. You feed that tree into the rest of the broccoli-jshint pipeline.

Worth noting: we're investigating moving to ESLint by default for the Ember ecosystem. Regardless of anything that comes out of that plan we'll continue to support JSHint for a long while still.

**Comment by rwjblue** [comment]
Can you revert the formatting only changes?  Specifically the `function(stuff) { }` to `function (stuff) { }` ones, they make it seem like many more changes are happening here, than actually are (and I personally prefer the non-space version).

File: index.js [hunk 1]
```diff
@@ -29,18 +34,26 @@ function JSHinter (inputNode, options) {
...
 JSHinter.prototype.extensions = ['js'];
 JSHinter.prototype.targetExtension = 'lint.js';
 
-JSHinter.prototype.baseDir = function() {
+JSHinter.prototype.baseDir = function () {
   return __dirname;
 };
 
+JSHinter.prototype.processIgnoredFiles = function () {
+  var funnel = Remove(this.outputPath, {
```
Comment (rwjblue) [inline index.js]: Where does `Remove` come from?

File: index.js [hunk 2]
```diff
@@ -29,18 +34,26 @@ function JSHinter (inputNode, options) {
...
+    paths: this.ignoredFiles || []
+  });
+  var builder = new broccoli.Builder(funnel);
```
Comment (rwjblue) [inline index.js]: We should not be invoking the `broccoli.Builder` in the plugin itself.  Also, I do not believe that `broccoli` is imported (it is a devDep only).

File: index.js [hunk 3]
```diff
@@ -29,18 +34,26 @@ function JSHinter (inputNode, options) {
...
 
 JSHinter.prototype.extensions = ['js'];
 JSHinter.prototype.targetExtension = 'lint.js';
 
-JSHinter.prototype.baseDir = function() {
+JSHinter.prototype.baseDir = function () {
   return __dirname;
 };
 
+JSHinter.prototype.processIgnoredFiles = function () {
```
Comment (rwjblue) [inline index.js]: Actually, it seems that this function is completely unused, can you remove it?

**Comment by vitch** [comment]
What's the status of this PR? Is there anything outstanding before it can be merged? I was looking for exactly this functionality...

**Comment by nathanhammond** [comment]
Our current plans include moving to eslint so this is low priority. Still want to do smart things with trees here.

**Pull request closed without merging**